ion-schema/src/constraint.rs

```diff
@@ -1031,6 +1067,117 @@ impl ConstraintValidator for FieldsConstraint {
     }
 }
 
+/// Implements an `field_names` constraint of Ion Schema
+/// [field_names]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#field_names
+#[derive(Debug, Clone, PartialEq)]
+pub struct FieldNamesConstraint {
+    type_reference: TypeReference,
+    requires_distinct: bool,
+}
+
+impl FieldNamesConstraint {
+    pub fn new(type_reference: TypeReference, requires_distinct: bool) -> Self {
+        Self {
+            type_reference,
+            requires_distinct,
+        }
+    }
+}
+
+impl ConstraintValidator for FieldNamesConstraint {
+    fn validate(
+        &self,
+        value: &IonSchemaElement,
+        type_store: &TypeStore,
+        ion_path: &mut IonPath,
+    ) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // create a set for checking duplicate elements
+        let mut element_set = vec![];
```

Comment:
Both `Vec` and `HashSet` have amortized `O(1)` time for inserting an element (assuming at the end, in the case of `Vec`). However, `Vec` has `O(n)` time complexity for checking if a single value is present in an unsorted list, so checking for duplicates is `O(n^2)`, but the `HashSet`, has `O(1)` time complexity to check for membership in the set, so checking for duplicates is overall (amortized) `O(n)`.

Reply:
I am using `Vec` as `Hash` implementation is not available for `Element`.

Reply:
But `Hash` is available for `String`, and we can convert the field names to `String`.

Reply:
Thats true :)